Pd 1057 new content for expanding a raidz storage pool (#2931)
truenas/documentation: content/SCALE/SCALETutorials/Storage/ManagePoolsScale.md

You can take a disk offline, detach it, replace it, manage the SED encryption password, and perform other disk management tasks from this screen.

See [Replacing Disks]({{< relref "ReplacingDisks.md" >}}) for more information on the **Offline**, **Replace** and **Online** options.

## Expanding a Pool

You can increase the size of an existing pool by adding one or more drives to an existing RAIDZ VDEV, adding a new VDEV of the same type, or by replacing all existing disks in the VDEV with larger disks.

By default, a VDEV limits all disks to the usable capacity of the smallest attached device.
TrueNAS SCALE automatically expands the usable capacity of the pool to fit all available space once the smallest attached disk is replaced (see [Replacing Disks to Expand a Pool](#replacing-disks-to-expand-a-pool) below).
If a pool is not automatically expanded, for example when resizing virtual disks in a hypervisor apart from TrueNAS, click **Expand** on the **Storage Dashboard** to manually increase the pool size to match all available disk space.

### Extending a RAIDZ VDEV

Additional disks can be added one at a time to a RAIDZ VDEV, expanding its capacity incrementally.

> This can be useful for small pools (typically with only one RAID-Z VDEV), where there isn't sufficient hardware capacity to add a mirrored VDEV by doubling the number of disks.

Reviewer: ```suggestion
This can be useful for small pools (typically with only one RAID-Z VDEV), where there isn't sufficient hardware capacity to add a second VDEV, doubling the number of disks.
```

{{< expand "Overview and Considerations" "v" >}}
TrueNAS SCALE 24.10 (Electric Eel) introduces RAIDZ extension to allow incremental expansion of an existing RAIDZ VDEV using one more disks.
RAIDZ extension allows resource- or hardware-limited homelab and small enterprise users to expand storage capacity with lower upfront costs compared to traditional ZFS expansion methods.

To expand a RAIDZ array, TrueNAS reads data from the current disks and rewrites it onto the new configuration, including any additional disks.

Data redundancy is maintained.

> If a disk fails mid-expansion, the process pauses until the RAIDZ virtual device (vdev) is healthy again, typically by replacing the failed disk and waiting for the system to rebuild.

Reviewer: Lets first say that the pool must be healthy to start expansion.

Author: ```suggestion
Make sure the pool is healthy before beginning the expansion process.
If a disk fails mid-expansion, the process pauses until the RAIDZ virtual device (vdev) is healthy again, typically by replacing the failed disk and waiting for the system to rebuild.
```

{{< trueimage src="/images/SCALE/Storage/DevicesExtend.png" alt="Devices Screen" id="Devices Screen" >}}

Select the target VDEV and click **Extend** to open the **Extend Vdev** window.

{{< trueimage src="/images/SCALE/Storage/ExtendVdev.png" alt="Extend Vdev" id="Extend Vdev" >}}

Select an available disk from the **New Disk** dropdown menu.
Click **Extend**.

A job progress window appears.
TrueNAS SCALE returns to the ***Poolname* Devices** screen when complete.

### Adding a VDEV to a Pool

ZFS supports adding VDEVs to an existing ZFS pool to increase the capacity or performance of the pool.
To extend a pool by mirroring, you must add a data VDEV of the same type as existing VDEVs.

{{< hint type=tip >}}
You cannot change the original encryption or data VDEV configuration.
{{< /hint >}}

{{< expand "Adding VDEV Examples" "v" >}}
* To make a striped mirror, add the same number of drives to extend a ZFS mirror.
  For example, you start with ten available drives. Begin by creating a mirror of two drives, and then extending the mirror by adding another mirror of two drives. Repeat this three more times until you add all ten drives.

> * To make a stripe of two RAIDZ1 VDEVs (similar to RAID 50 on a hardware controller), add another three drives to extend the three-drive RAIDZ1.
> * To make a stripe of RAIDZ2 VDEVs (similar to RAID 60 on a hardware controller), add another four drives to extend the four-drive RAIDZ2.

Reviewer: ```suggestion
* To make a stripe of two 3-drive RAIDZ1 VDEVs (similar to RAID 50 on a hardware controller), add another three drives as a new RAIDZ1 VDEV to existing single 3-drive RAIDZ1 VDEV pool.
* To make a stripe of two 6-disk RAIDZ2 VDEVs (similar to RAID 60 on a hardware controller), add another six drives as a new RAIDZ2 VDEV to existing single 6-drive RAIDZ2 VDEV pool.
```
I'd propose to avoid word "extend" here to not confuse with RAIDZ expansion.  Also 4-disk RAIDZ2 is not the best configuration example, so I'd change it to 6 (or 10, if you like).

Click on a disk to see the devices widgets for that disk.
You can take a disk offline, detach it, replace it, manage the SED encryption password, and perform other disk management tasks from this screen.

See [Replacing Disks]({{< relref "ReplacingDisks.md" >}}) for more information on the **Offline**, **Replace** and **Online** options.

## Expanding a Pool

You can increase the size of an existing pool by adding one or more drives to an existing RAIDZ VDEV, adding a new VDEV of the same type, or by replacing all existing disks in the VDEV with larger disks.

By default, a VDEV limits all disks to the usable capacity of the smallest attached device.
TrueNAS SCALE automatically expands the usable capacity of the pool to fit all available space once the smallest attached disk is replaced (see [Replacing Disks to Expand a Pool](#replacing-disks-to-expand-a-pool) below).
If a pool is not automatically expanded, for example when resizing virtual disks in a hypervisor apart from TrueNAS, click **Expand** on the **Storage Dashboard** to manually increase the pool size to match all available disk space.

### Extending a RAIDZ VDEV

> Additional disks can be added one at a time to a RAIDZ VDEV, expanding its capacity incrementally.

Author: ```suggestion
Extend a RAIDZ VDEV to add additional disks one at a time, expanding capacity incrementally.
```

The storage pool remains accessible throughout the expansion.
If you reboot or export/import the pool, the expansion resumes from where it left off.

After the expansion, the extra space becomes available for use.

The fault tolerance level of the RAIDZ array remains unchanged.
For example, a RAIDZ2 stays a RAIDZ2 even after adding more disks.

You can expand a RAIDZ vdev multiple times.

Existing data blocks retain their original data-to-parity ratio and block width, but spread across the larger set of disks.
New data blocks adopt the new data-to-parity ratio and width.
Because of this overhead, an extended RAIDZ VDEV can report a lower total capacity than a freshly created VDEV with the same number of disks.

{{< trueimage src="/images/Reference/RaidzExpansion.png" alt="RAIDZ Expansion" id="RAIDZ Expansion" caption="Before (left) and after (right) expansion of a four-disk to five-disk RAIDZ1<br>Thanks to Matt Ahrens ([Source](https://arstechnica.com/gadgets/2021/06/raidz-expansion-code-lands-in-openzfs-master/))" >}}

> The additional overhead of now-undersized existing blocks can be recovered by reading and rewriting existing data to the extended pool.
> This can occur naturally over the lifetime of the pool as ZFS rewrites data blocks when you modify or delete data.
> Alternately, you can manually replicate and rewrite the data to the pool.
> Data is rewritten using the extended block size and parity ratio.

Author: ```suggestion
Extended VDEVs recover lost headroom as existing data is read and rewritten to the new parity ratio.
This can occur naturally over the lifetime of the pool as you modify or delete data.
To manually recover capacity, simply replicate and rewrite the data to the extended pool.
```

You can  use the [RAIDZ Extension Calculator]({{< relref "/References/ExtensionCalculator.md" >}}) to visualize potential lost headroom and capacity available to recover by rewriting existing data.

For more information, see [Jim Salter's article](https://arstechnica.com/gadgets/2021/06/raidz-expansion-code-lands-in-openzfs-master/) at Ars Technica and the upstream [RAIDZ extension](https://github.com/openzfs/zfs/pull/15022) PR, sponsored by iXsystems, at OpenZFS.

> See also Louwrentius' ["ZFS RAIDZ Expansion Is Awesome but Has a Small Caveat"](https://louwrentius.com/zfs-raidz-expansion-is-awesome-but-has-a-small-caveat.html) for an in-depth discussion of lost capacity and recovering overhead.

Author: ```suggestion
See also ["ZFS RAIDZ Expansion Is Awesome but Has a Small Caveat"](https://louwrentius.com/zfs-raidz-expansion-is-awesome-but-has-a-small-caveat.html) by Louwrentius for an in-depth discussion of lost capacity and recovering overhead.
```

To save changes click **Update Pool**.

### Replacing Disks to Expand a Pool

> To expand a pool by replacing disks, you must individually offline each disk in the pool and replace it with a higher capacity disk following the same procedure as in [Replacing Disks]({{< relref "ReplacingDisks.md" >}}).

Reviewer: Offlining the disk should be required only if you have to replace the disks in place.  The downside of it is reduced redundancy during the process.  Safer should be to insert the new disk into an empty enclosure slot and remove the old disk only after replace operation completed.

Author: ```suggestion
To expand a pool by replacing disks, you must individually replace each with a higher capacity disk following the same procedure as in [Replacing Disks]({{< relref "ReplacingDisks.md" >}}).
```

Reviewer: It seem the procedure also documents only the in-place replace.  We may wish to test and document other approach too.

1. Go to the **Storage Dashboard** and click **Manage Devices** on the **Topology** widget for the pool to open the ***Poolname* Devices** screen.
Click anywhere on the VDEV to expand it and select one of the existing disks.

> 2. Take the disk offline.

Author: ```suggestion
2. (Optional) If replacing disks in place, take one existing disk offline.
```

   {{< trueimage src="/images/SCALE/Storage/DevicesDiskWidgets.png" alt="Devices Disk Widgets" id="Devices Disk Widgets" >}}

>    Click **Offline** on the **ZFS Info** widget to take the disk offline. The button toggles to **Online**.

Author: ```suggestion
   Click **Offline** on the **ZFS Info** widget to take the disk offline. The button toggles to **Online**.
   
   Remove the disk from the system.
```

> 3. Remove the disk from the system and replace it with a larger capacity disk.

Author: ```suggestion
3. Insert a larger capacity disk into an open enclosure slot.
```
